types/lru/lru.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package lru contains a typed Least-Recently-Used cache.
+package lru
+
+import (
+	"container/list"
+)
+
+// Cache is an LRU cache keyed by K, storing V.
+//
+// The zero value is valid to use.
+//
+// It is not safe for concurrent access.
+type Cache[K comparable, V any] struct {
+	// MaxEntries is the maximum number of cache entries before
+	// an item is evicted. Zero means no limit.
+	MaxEntries int
```

Comment:
I've had need for caches that we're evicted by time rather than MaxEntries, but that can be a future consideration.